In [27]:
import pandas as pd
import numpy as np


[Dados de emissao de CO2 por país para a análise](https://www.kaggle.com/datasets/moazzimalibhatti/co2-emission-by-countries-year-wise-17502022/data)

In [28]:
df = pd.read_csv("CO2 emission by countries.csv",sep=",",encoding='unicode_escape')
display(df)

Country Code Calling Code  Year  CO2 emission (Tons)  \
0      Afghanistan   AF           93  1750                  0.0   
1      Afghanistan   AF           93  1751                  0.0   
2      Afghanistan   AF           93  1752                  0.0   
3      Afghanistan   AF           93  1753                  0.0   
4      Afghanistan   AF           93  1754                  0.0   
...            ...  ...          ...   ...                  ...   
59615     Zimbabwe   ZW          263  2016          736467042.0   
59616     Zimbabwe   ZW          263  2017          746048675.0   
59617     Zimbabwe   ZW          263  2018          757903042.0   
59618     Zimbabwe   ZW          263  2019          768852126.0   
59619     Zimbabwe   ZW          263  2020          779383468.0   

       Population(2022)      Area % of World Density(km2)  
0            41128771.0  652230.0      0.40%       63/km²  
1            41128771.0  652230.0      0.40%       63/km²  
2            41128771.0  652230.0      0.40%       63/km²  
3            41128771.0  652230.0      0.40%       63/km²  
4            41128771.0  652230.0      0.40%       63/km²  
...                 ...       ...        ...          ...  
59615        16320537.0  390757.0      0.30%       42/km²  
59616        16320537.0  390757.0      0.30%       42/km²  
59617        16320537.0  390757.0      0.30%       42/km²  
59618        16320537.0  390757.0      0.30%       42/km²  
59619        16320537.0  390757.0      0.30%       42/km²  

[59620 rows x 9 columns]

In [29]:
 # Pega amostra de todas as colunas
df_sample = df.head(20)
df_sample.to_csv('sampleH.csv')
df_sample

Country Code Calling Code  Year  CO2 emission (Tons)  \
0   Afghanistan   AF           93  1750                  0.0   
1   Afghanistan   AF           93  1751                  0.0   
2   Afghanistan   AF           93  1752                  0.0   
3   Afghanistan   AF           93  1753                  0.0   
4   Afghanistan   AF           93  1754                  0.0   
5   Afghanistan   AF           93  1755                  0.0   
6   Afghanistan   AF           93  1756                  0.0   
7   Afghanistan   AF           93  1757                  0.0   
8   Afghanistan   AF           93  1758                  0.0   
9   Afghanistan   AF           93  1759                  0.0   
10  Afghanistan   AF           93  1760                  0.0   
11  Afghanistan   AF           93  1761                  0.0   
12  Afghanistan   AF           93  1762                  0.0   
13  Afghanistan   AF           93  1763                  0.0   
14  Afghanistan   AF           93  1764                  0.0   
15  Afghanistan   AF           93  1765                  0.0   
16  Afghanistan   AF           93  1766                  0.0   
17  Afghanistan   AF           93  1767                  0.0   
18  Afghanistan   AF           93  1768                  0.0   
19  Afghanistan   AF           93  1769                  0.0   

    Population(2022)      Area % of World Density(km2)  
0         41128771.0  652230.0      0.40%       63/km²  
1         41128771.0  652230.0      0.40%       63/km²  
2         41128771.0  652230.0      0.40%       63/km²  
3         41128771.0  652230.0      0.40%       63/km²  
4         41128771.0  652230.0      0.40%       63/km²  
5         41128771.0  652230.0      0.40%       63/km²  
6         41128771.0  652230.0      0.40%       63/km²  
7         41128771.0  652230.0      0.40%       63/km²  
8         41128771.0  652230.0      0.40%       63/km²  
9         41128771.0  652230.0      0.40%       63/km²  
10        41128771.0  652230.0      0.40%       63/km²  
11        41128771.0  652230.0      0.40%       63/km²  
12        41128771.0  652230.0      0.40%       63/km²  
13        41128771.0  652230.0      0.40%       63/km²  
14        41128771.0  652230.0      0.40%       63/km²  
15        41128771.0  652230.0      0.40%       63/km²  
16        41128771.0  652230.0      0.40%       63/km²  
17        41128771.0  652230.0      0.40%       63/km²  
18        41128771.0  652230.0      0.40%       63/km²  
19        41128771.0  652230.0      0.40%       63/km²

In [30]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59620 entries, 0 to 59619
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Country              59620 non-null  object 
 1   Code                 57452 non-null  object 
 2   Calling Code         56097 non-null  object 
 3   Year                 59620 non-null  int64  
 4   CO2 emission (Tons)  59620 non-null  float64
 5   Population(2022)     53116 non-null  float64
 6   Area                 55284 non-null  float64
 7   % of World           55284 non-null  object 
 8   Density(km2)         53116 non-null  object 
dtypes: float64(3), int64(1), object(5)
memory usage: 4.1+ MB


In [31]:
df = df.drop(["Calling Code","Code",'% of World',"Area"], axis=1)

In [32]:
display(df)

Country  Year  CO2 emission (Tons)  Population(2022) Density(km2)
0      Afghanistan  1750                  0.0        41128771.0       63/km²
1      Afghanistan  1751                  0.0        41128771.0       63/km²
2      Afghanistan  1752                  0.0        41128771.0       63/km²
3      Afghanistan  1753                  0.0        41128771.0       63/km²
4      Afghanistan  1754                  0.0        41128771.0       63/km²
...            ...   ...                  ...               ...          ...
59615     Zimbabwe  2016          736467042.0        16320537.0       42/km²
59616     Zimbabwe  2017          746048675.0        16320537.0       42/km²
59617     Zimbabwe  2018          757903042.0        16320537.0       42/km²
59618     Zimbabwe  2019          768852126.0        16320537.0       42/km²
59619     Zimbabwe  2020          779383468.0        16320537.0       42/km²

[59620 rows x 5 columns]

In [33]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59620 entries, 0 to 59619
Data columns (total 5 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Country              59620 non-null  object 
 1   Year                 59620 non-null  int64  
 2   CO2 emission (Tons)  59620 non-null  float64
 3   Population(2022)     53116 non-null  float64
 4   Density(km2)         53116 non-null  object 
dtypes: float64(2), int64(1), object(2)
memory usage: 2.3+ MB


In [34]:
df.head()

Country  Year  CO2 emission (Tons)  Population(2022) Density(km2)
0  Afghanistan  1750                  0.0        41128771.0       63/km²
1  Afghanistan  1751                  0.0        41128771.0       63/km²
2  Afghanistan  1752                  0.0        41128771.0       63/km²
3  Afghanistan  1753                  0.0        41128771.0       63/km²
4  Afghanistan  1754                  0.0        41128771.0       63/km²

In [35]:
df = df.groupby('Country').sum()


<ipython-input-35-295f1f8e805a>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  df = df.groupby('Country').sum()


In [37]:
df = df.drop(["Year","Population(2022)"], axis=1)

In [40]:
df['CO2 emission (Tons)'] = df['CO2 emission (Tons)'].map('{:.0f}'.format)
df

CO2 emission (Tons)
Country                              
Afghanistan                3754106020
Albania                    9250245605
Algeria                   95820242814
Andorra                     236072246
Angola                    11431577776
...                               ...
Vietnam                   72625031431
Wallis and Futuna             5293220
Yemen                     14217319264
Zambia                     8811258531
Zimbabwe                  29364694233

[220 rows x 1 columns]

In [55]:
display(df)

CO2 emission (Tons)
Country                              
Afghanistan                3754106020
Albania                    9250245605
Algeria                   95820242814
Andorra                     236072246
Angola                    11431577776
...                               ...
Vietnam                   72625031431
Wallis and Futuna             5293220
Yemen                     14217319264
Zambia                     8811258531
Zimbabwe                  29364694233

[220 rows x 1 columns]

In [43]:
df.loc["United States"]

CO2 emission (Tons)    19068767447823
Name: United States, dtype: object

In [45]:
df_2= df.copy()

In [53]:
# Converte coluna para float
df_2["CO2 emission (Tons)"] = df_2["CO2 emission (Tons)"].astype(float)

# Agora aplica a formatação
df_2["CO2 col_formatada"] = df_2["CO2 emission (Tons)"].apply(lambda x: '{:,.2f}'.format(x).replace(',', '.'))

# **Emissão média de cada pais ao longo de 272 anos **
### Diferença de tempo entre inicio da revolução industrial, 1750, até o ano dos dados 2022

In [63]:
df_2["Media de emissão anual de CO2"] = df_2["CO2 emission (Tons)"]/272

In [70]:
display(df_2)
#df_2.info()

CO2 emission (Tons)  CO2 col_formatada  \
Country                                                     
Afghanistan               3.754106e+09   3.754.106.020.00   
Albania                   9.250246e+09   9.250.245.605.00   
Algeria                   9.582024e+10  95.820.242.814.00   
Andorra                   2.360722e+08     236.072.246.00   
Angola                    1.143158e+10  11.431.577.776.00   
...                                ...                ...   
Vietnam                   7.262503e+10  72.625.031.431.00   
Wallis and Futuna         5.293220e+06       5.293.220.00   
Yemen                     1.421732e+10  14.217.319.264.00   
Zambia                    8.811259e+09   8.811.258.531.00   
Zimbabwe                  2.936469e+10  29.364.694.233.00   

                   Media de emissão anual de CO2 Media formatada  
Country                                                           
Afghanistan                         1.380186e+07   13,801,860.37  
Albania                             3.400826e+07   34,008,255.90  
Algeria                             3.522803e+08  352,280,304.46  
Andorra                             8.679127e+05      867,912.67  
Angola                              4.202786e+07   42,027,859.47  
...                                          ...             ...  
Vietnam                             2.670038e+08  267,003,792.03  
Wallis and Futuna                   1.946037e+04       19,460.37  
Yemen                               5.226956e+07   52,269,556.12  
Zambia                              3.239433e+07   32,394,332.83  
Zimbabwe                            1.079584e+08  107,958,434.68  

[220 rows x 4 columns]

In [65]:
df_2["Media formatada"] = df_2["Media de emissão anual de CO2"].apply(lambda x: '{:,.2f}'.format(x))#.replace(',', '.'))
display(df_2)

CO2 emission (Tons)  CO2 col_formatada  \
Country                                                     
Afghanistan               3.754106e+09   3.754.106.020.00   
Albania                   9.250246e+09   9.250.245.605.00   
Algeria                   9.582024e+10  95.820.242.814.00   
Andorra                   2.360722e+08     236.072.246.00   
Angola                    1.143158e+10  11.431.577.776.00   
...                                ...                ...   
Vietnam                   7.262503e+10  72.625.031.431.00   
Wallis and Futuna         5.293220e+06       5.293.220.00   
Yemen                     1.421732e+10  14.217.319.264.00   
Zambia                    8.811259e+09   8.811.258.531.00   
Zimbabwe                  2.936469e+10  29.364.694.233.00   

                   Media de emissão anual de CO2 Media formatada  
Country                                                           
Afghanistan                         1.380186e+07   13,801,860.37  
Albania                             3.400826e+07   34,008,255.90  
Algeria                             3.522803e+08  352,280,304.46  
Andorra                             8.679127e+05      867,912.67  
Angola                              4.202786e+07   42,027,859.47  
...                                          ...             ...  
Vietnam                             2.670038e+08  267,003,792.03  
Wallis and Futuna                   1.946037e+04       19,460.37  
Yemen                               5.226956e+07   52,269,556.12  
Zambia                              3.239433e+07   32,394,332.83  
Zimbabwe                            1.079584e+08  107,958,434.68  

[220 rows x 4 columns]

#Ranking de emissão média anul de CO2 por pais

In [69]:
df_ordenado = df_2.copy()

In [71]:
df_ordenado = df_ordenado.sort_values(by='Media de emissão anual de CO2', ascending=False)

In [72]:
display(df_ordenado)

CO2 emission (Tons)      CO2 col_formatada  \
Country                                                         
United States             1.906877e+13  19.068.767.447.823.00   
United Kingdom            6.162082e+12   6.162.081.873.517.00   
Germany                   5.300531e+12   5.300.530.860.522.00   
China                     4.162896e+12   4.162.896.307.149.00   
Russia                    4.080739e+12   4.080.738.685.231.00   
...                                ...                    ...   
Saint Helena              6.904616e+06           6.904.616.00   
Niue                      6.347078e+06           6.347.078.00   
Wallis and Futuna         5.293220e+06           5.293.220.00   
Tuvalu                    4.456110e+06           4.456.110.00   
Antarctica                3.843536e+06           3.843.536.00   

                   Media de emissão anual de CO2    Media formatada  
Country                                                              
United States                       7.010576e+10  70,105,762,675.82  
United Kingdom                      2.265471e+10  22,654,712,770.28  
Germany                             1.948725e+10  19,487,245,810.74  
China                               1.530477e+10  15,304,765,835.11  
Russia                              1.500272e+10  15,002,715,754.53  
...                                          ...                ...  
Saint Helena                        2.538462e+04          25,384.62  
Niue                                2.333485e+04          23,334.85  
Wallis and Futuna                   1.946037e+04          19,460.37  
Tuvalu                              1.638276e+04          16,382.76  
Antarctica                          1.413065e+04          14,130.65  

[220 rows x 4 columns]